File: backend/app/models/task.py
"""Task domain models.

Defines database representations for tasks and executions.
"""

from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class TaskStatus(StrEnum):
    """Execution status for a task."""

    PENDING = "PENDING"
    PLANNING = "PLANNING"
    RUNNING = "RUNNING"
    WAITING = "WAITING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


class TaskPriority(StrEnum):
    """Priority level for a task."""

    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    URGENT = "URGENT"


class Task(BaseModel):
    """Task database representation model."""

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        use_enum_values=True,
    )

    id: str | None = Field(default=None, alias="id")
    user_id: str
    title: str
    description: str | None = None
    goal: str
    status: TaskStatus = TaskStatus.PENDING
    priority: TaskPriority = TaskPriority.MEDIUM
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    started_at: datetime | None = None
    completed_at: datetime | None = None
    deleted_at: datetime | None = None


class TaskExecution(BaseModel):
    """Domain model tracking a single execution run attempt of a task."""

    model_config = ConfigDict(
        populate_by_name=True,
        arbitrary_types_allowed=True,
        use_enum_values=True,
    )

    id: str | None = Field(default=None, alias="id")
    task_id: str
    status: TaskStatus
    started_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: datetime | None = None
    error: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)

File: frontend/src/types/task.ts
export enum TaskStatus {
  PENDING = "PENDING",
  PLANNING = "PLANNING",
  READY = "READY",
  RUNNING = "RUNNING",
  WAITING = "WAITING",
  COMPLETED = "COMPLETED",
  FAILED = "FAILED",
  CANCELLED = "CANCELLED",
}

export enum TaskPriority {
  LOW = "LOW",
  MEDIUM = "MEDIUM",
  HIGH = "HIGH",
  URGENT = "URGENT",
}

export interface Task {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  goal: string;
  status: TaskStatus;
  priority: TaskPriority;
  metadata: Record<string, any>;
  created_at: string;
  updated_at: string;
  started_at: string | null;
  completed_at: string | null;
}

export interface TaskCreate {
  title: string;
  description?: string | null;
  goal: string;
  priority?: TaskPriority;
  metadata?: Record<string, any>;
}

export interface TaskUpdate {
  title?: string;
  description?: string | null;
  goal?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  metadata?: Record<string, any>;
}
